File: tests/mv-2c/duration.py
#! /bin/env python
import math
from typing import Callable
import re2 as re
from json_model.runtime import *
__version__ = "2"

def check_model_fun(name: str) -> CheckFun:
    """Return check function for JSON model name."""
    return check_model_map[name]

# entry point for generated checkers
def check_model(val: Jsonable, name: str = "", rep: Report = None) -> bool:
    """Check val validity against JSON Model name."""
    if name not in check_model_map:
        raise Exception(f"unexpected model name: {name}")
    checker = check_model_map[name]
    return checker(val, [], rep)

check_model_map: PropMap
jm_is_duration_reco: object
jm_is_duration: RegexFun

def json_model_1(val: Jsonable, path: Path, rep: Report) -> bool:
    return isinstance(val, str) and jm_is_duration(val, None, None)



# initialization guard
initialized: bool = False

# differed module initializations
def check_model_init():
    global initialized
    if not initialized:
        initialized = True
        global check_model_map
        check_model_map = {
            "": json_model_1,
        }
        global jm_is_duration_reco, jm_is_duration
        jm_is_duration_reco = re.compile("^P(([0-9]+D|[0-9]+M([0-9]+D)?|[0-9]+Y([0-9]+M([0-9]+D)?)?)(T([0-9]+H([0-9]+M([0-9]+S)?)?|[0-9]+M([0-9]+S)?|[0-9]+S))?|T([0-9]+H([0-9]+M([0-9]+S)?)?|[0-9]+M([0-9]+S)?|[0-9]+S)|[0-9]+W)$")
        jm_is_duration = lambda s, p, r: jm_is_duration_reco.search(s) is not None

# differed module cleanup
def check_model_free():
    global initialized
    if initialized:
        initialized = False
        global jm_is_duration_reco, jm_is_duration
        jm_is_duration_reco = None
        jm_is_duration = None

if __name__ == "__main__":
    check_model_init()
    main(check_model_fun, check_model_map, __version__)
    check_model_free()

File: tests/mv-2c/duration.java

import json_model.*;
import java.util.Map;
import java.util.Set;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.regex.Pattern;
import java.util.regex.Matcher;

@SuppressWarnings("unchecked")
public class duration extends ModelChecker
{
    static public final String VERSION = "2";

    public Map<String, Checker> duration_map_pmap;
    public Pattern jm_is_duration_pat = null;

    public boolean json_model_1(Object val, Path path, Report rep)
    {
        return json.isString(val) && jm_is_duration(json.asString(val), null, null);
    }


    public boolean jm_is_duration(String val, Path path, Report rep)
    {
        return jm_is_duration_pat.matcher(val).find();
    }

    public void init(JSON json)
    {
        if (!initialized)
        {
            try {
            duration_map_pmap = new HashMap<String, Checker>();
            duration_map_pmap.put("", new Checker() { public boolean call(Object o, Path p, Report r) { return json_model_1(o, p, r);} });
            jm_is_duration_pat = Pattern.compile("^P(([0-9]+D|[0-9]+M([0-9]+D)?|[0-9]+Y([0-9]+M([0-9]+D)?)?)(T([0-9]+H([0-9]+M([0-9]+S)?)?|[0-9]+M([0-9]+S)?|[0-9]+S))?|T([0-9]+H([0-9]+M([0-9]+S)?)?|[0-9]+M([0-9]+S)?|[0-9]+S)|[0-9]+W)$");
                super.init(json);
            }
            catch (Exception e) {
                throw new Error("cannot initialized model checker: " + e);
            }
        }
    }

    public void free()
    {
        if (initialized)
        {
            super.free();
            duration_map_pmap = null;
            jm_is_duration_pat = null;
        }
    }

    public Checker get(String name)
    {
        return duration_map_pmap.get(name);
    }

    public Set<String> models()
    {
        return duration_map_pmap.keySet();
    }

    static public void main(String[] args) throws Exception
    {
        ModelChecker checker = new duration();
        Main.main("duration", checker, VERSION, args);
    }
}
